File: Number of Turns in Binary Tree.cpp
class Solution{
  public:
  
    // function should return the number of turns required to go from first node to second node 
    int NumberOFTurns(struct Node* root, int first, int second)
    {
      // Your code goes here
      Node *head = lca(root, first, second);
      
      int F = countTurns(head, first, ' ');
      int S = countTurns(head, second, ' ');
      
      int ans = F + S;
      
      if(head->data != first and head->data != second)
      ans++;
      
      return ans;
    }
    
    //Function to return the lowest common ancestor in a Binary Tree.
    Node* lca(Node* root ,int n1 ,int n2 )
    {
       //Your code here 
       
        if(!root)
        return root;
        
        if(root->data == n1 or root->data == n2)
        return root;
        
        Node *Left = lca(root->left, n1, n2);
        Node *Right = lca(root->right, n1, n2);
        
        if(Left and !Right)
        return Left;
        
        else if(!Left and Right)
        return Right;
        
        else if(Left and Right)
        return root;
        
        else
        return nullptr;
    }
    
    // function should return the minimum number of turn to reach a node.
    int countTurns(Node* root, int n, char d)
    {
        if(!root)
        return 1e5;
        
        if(root->data == n)
        return 0;
        
        int Left = countTurns(root->left, n, 'L');
        int Right = countTurns(root->right, n, 'R');
        
        if(d == 'L')
        Right++;
        if(d == 'R')
        Left++;
        
        return min(Left, Right);
    }
};

File: Number of Turns in Binary Tree.java
class Solution
{
    static int ans;
    static int NumberOfTurns(Node root, int first, int second)
    {
        //your code here
        ans=0;
        Node lca=findLCA(root,first,second);
        boolean left=true;
        if(lca.data==first){
            findCount(lca.left,left,second);
            findCount(lca.right,!left,second);
        }else if(lca.data==second){
            findCount(lca.left,left,first);
            findCount(lca.right,!left,first);
        }else{
            findCount(lca.left,left,first);
            findCount(lca.right,!left,first);
            findCount(lca.left,left,second);
            findCount(lca.right,!left,second);
            return ans+1;
        }
        return ans;
    }
    
    static boolean findCount(Node node,boolean left,int key){
        if(node==null)return false;
        if(node.data==key)return true;
        
        if(left){
            if(findCount(node.left,left,key)){
                return true;
            }
            if(findCount(node.right,!left,key)){
                ans++;
                return true;
            }
        }else{
            if(findCount(node.right,left,key)){
                return true;
            }
            if(findCount(node.left,!left,key)){
                ans++;
                return true;
            }
        }
        return false;
    }
    
    static Node findLCA(Node root,int first,int second){
        if(root==null)return null;
        if(root.data==first || root.data==second){
            return root;
        }
        
        Node left_lca=findLCA(root.left,first,second);
        Node right_lca=findLCA(root.right,first,second);
        if(left_lca!=null && right_lca!=null){
            return root;
        }else{
            return left_lca==null?right_lca:left_lca;
        }
    }
}
